File: java/L300/LeetCode0318.java
package L300;

public class LeetCode0318 {
    public int maxProduct(String[] words) {
        int[] bits = new int[words.length];
        for (int i = 0; i < words.length; i++) {
            int bit = 0;
            for (char c : words[i].toCharArray()) {
                bit |= 1 << (c - 'a');
            }
            bits[i] |= bit;
        }
        int max = 0;
        for (int i = 0; i < words.length; i++) {
            for (int j = i + 1; j < words.length; j++) {
                if ((bits[i] & bits[j]) == 0) {
                    max = Math.max(max, words[i].length() * words[j].length());
                }
            }
        }
        return max;
    }
}

File: cpp/L300/LeetCode0318.cc
#include <utility>
#include <vector>
using std::vector;
#include <string>
using std::string;

class Solution {
public:
    int maxProduct(vector<string>& words) {
        int N = words.size();
        vector<std::pair<int, int>> vp(N);
        for (int i = 0;i < N;++i) {
            for (char const c : words[i]) {
                vp[i].first |= (1 << (c - 'a'));
            }
            vp[i].second = words[i].size();
        }
        int ans{0};
        for (int i = 0;i < N;++i) {
            for (int j{i + 1};j < N;++j) {
                if ((vp[i].first & vp[j].first) == 0) {
                    ans = std::max(ans, vp[i].second * vp[j].second);
                }
            }
        }
        return ans;
    }
};